### PR TITLE
#165305754 Fix validations  errors on signup

### PR DESCRIPTION
**Pull request by mdn-network**
#### Description
- Fix email validations error messages
- Fix password validation error messages
- Validates too common password
- Fix Username validation error messages
- Fix failing tests
#### Type of change
- [x] Bug fix (nonbreaking change which fixes an issue)
- [ ] New feature (nonbreaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

#### How Has This Been Tested?
Please describe the tests that you ran to verify your changes
- [ ] End to end
- [ ] Integration

#### Checklist:
- [x] My code follows the style guide for this project
- [x] I have linted my code prior to submission
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

#### Pivotal Tracker
[165305754](https://www.pivotaltracker.com/story/show/165305754)

### CODE REVIEW

**Review by Ogutu-Brian — CHANGES_REQUESTED** [review]
Hello @Philipotieno ,
You should make your pull request comment be in line with the development conventions in the Engineering playbook. 
The information you have provided here could be put in the description section. Instead of having
[Fixes  #165305754] in your PR, you could just use the story ID and specify that it is from the pivotal tracker, then add a link to the corresponding story.

File: authors/apps/authentication/tests/test_validations.py [hunk 1]
```diff
@@ -65,14 +65,4 @@ def test_short_password(self):
...
-    def test_non_alphanumeric_password(self):
-        """
-        Test user signup with the special characters as password
-        """
-        response = self.signup_user("test", "test@gmail.com", "@@@@@@@@@@")
-        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
-        errors = response.data['errors']
-        self.assertEqual(
-            errors['password'][0], "Password must have alphanumeric characters with no space")
+            errors['password'][1], "Password must have three lowercase, two lowercase, one special case letters and two digits eg 12aaaAA@")
```
Comment (Ogutu-Brian) [inline authors/apps/authentication/tests/test_validations.py]: Hello @Philipotieno ,
In this error message, you have the lowercase policy repeated. In addition, to make the system forgiving enough and password memorable, we could use this policy:
```
a. Minimum of 2 uppercase letters
b. Minimum of 2 special characters
c. Minimum of 2 digits
d. Minimum length of 8
```
With this, the password strengths would be strong and also memorable

**Review by Ogutu-Brian — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -27,7 +33,7 @@ class RegistrationSerializer(serializers.ModelSerializer):
             'required': 'Password is a required field',
             'min_length': 'Password must be at least 8 characters long',
             'max_length': 'Password cannot be more than 30 characters',
-            'invalid': 'Password must have alphanumeric characters with no space',
+            'invalid': 'Password must have three lowercase, two lowercase, one special case letters and two digits eg 12aaaAA@',
```
Comment (IssaIan) [inline authors/apps/authentication/serializers.py]: Please check this error message @Philipotieno

File: authors/apps/authentication/serializers.py [hunk 3]
```diff
@@ -45,25 +51,26 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
         validators=[
             UniqueValidator(
                 queryset=User.objects.all(),
-                message='username already taken',
+                message='user with this username already exists',
             )
         ],
         error_messages={
-            'invalid': 'Username should have no spaces or special characters only',
+            'invalid': 'Username should be at least 3 characters, have no spaces or special characters except ._-',
```
Comment (IssaIan) [inline authors/apps/authentication/serializers.py]: Also make the accepted special characters here clearer, right now one can't tell which are the accepted ones. Try naming the accepted ones and enclosing the character in a bracket as such ``hyphen (-), fullstop (.), e.t.c``

**Review by IssaIan — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/serializers.py (hunk 3 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: Hello @IssaIan ,
I think we could let our users choose any special character. That would help make their passwords unique. [special characters](https://www.owasp.org/index.php/Password_special_characters)

Comment (mdn-network) [reply authors/apps/authentication/serializers.py]: @Ogutu-Brian That's the username

File: authors/apps/authentication/tests/test_validations.py (hunk 1 above)
Comment (mdn-network) [reply authors/apps/authentication/tests/test_validations.py]: Ogutu I have corrected that

File: authors/apps/authentication/serializers.py (hunk 3 above)
Comment (Ogutu-Brian) [reply authors/apps/authentication/serializers.py]: I thought it was the password

File: authors/apps/authentication/serializers.py (hunk 2 above)
Comment (mdn-network) [reply authors/apps/authentication/serializers.py]: @IssaIan have resolved it

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -45,25 +51,26 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
         validators=[
             UniqueValidator(
                 queryset=User.objects.all(),
-                message='username already taken',
+                message='user with this username already exists',
             )
         ],
         error_messages={
-            'invalid': 'Username should have no spaces or special characters only',
+            'invalid': 'Username should be at least 3 characters, have no spaces or special characters except hyphen-, fullstop. and underscore_',
```
Comment (IssaIan) [inline authors/apps/authentication/serializers.py:72]: here the special characters acceptable are still confusing. I would suggest changing `hyphen-, fullstop. and underscore_'` to `hyphen (-), fullstop (.) and underscore (_)`

**Review by IssaIan — CHANGES_REQUESTED** [review]

**Review by Ogutu-Brian — APPROVED** [review]
LGTM

**Review by a-braham — APPROVED** [review]
LGTM

**Review by matthenge — APPROVED** [review]
LGTM

**Review by IssaIan — APPROVED** [review]

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (oluwagbenga-joloko) [reply authors/apps/authentication/serializers.py:72]: Issa's comment above is valid. Please apply the fixes suggested. @Philipotieno

**Pull request closed without merging**